src/test/kotlin/org/jitsi/nlj/stats/EndpointConnectionStatsTest.kt

```diff
@@ -0,0 +1,144 @@
+import io.kotlintest.shouldBe
+import io.kotlintest.specs.ShouldSpec
+import java.time.Duration
+import org.jitsi.nlj.resources.logging.StdoutLogger
+import org.jitsi.nlj.test_utils.FakeClock
+import org.jitsi.rtp.rtcp.RtcpReportBlock
+import org.jitsi.rtp.rtcp.RtcpRrPacket
+import org.jitsi.rtp.rtcp.RtcpSrPacket
+
+class EndpointConnectionStatsTest : ShouldSpec() {
+    override fun isolationMode(): IsolationMode? = IsolationMode.InstancePerLeaf
+
+    private val clock: FakeClock = spy()
+
+    private var mostRecentPublishedRtt: Double = -1.0
+    private var numRttUpdates: Int = 0
+    private val stats = EndpointConnectionStats(StdoutLogger(), clock).also {
+        it.addListener(object : EndpointConnectionStats.EndpointConnectionStatsListener {
+            override fun onRttUpdate(newRtt: Double) {
+                mostRecentPublishedRtt = newRtt
+                numRttUpdates++
+            }
+        })
+    }
+
+    init {
+        "after an SR is sent" {
+            val mockSenderInfo = mock<RtcpSrPacket.SenderInfo> {
+                on { compactedNtpTimestamp } doReturn 100
+            }
+            val srPacket = mock<RtcpSrPacket> {
+                on { senderSsrc } doReturn 1234L
+                on { senderInfo } doReturn mockSenderInfo
+            }
+
+            stats.rtcpPacketSent(srPacket)
+            "and an RR is received which refers to it" {
+                clock.elapse(60.milliseconds)
+
+                val reportBlock = mock<RtcpReportBlock> {
+                    on { ssrc } doReturn 1234L
+                    on { lastSrTimestamp } doReturn 100
+                    on { delaySinceLastSr } doReturn 50.milliseconds.toDelaySinceLastSrFraction()
+                }
+
+                val rtcpPacket = mock<RtcpRrPacket> {
+                    on { reportBlocks } doReturn listOf(reportBlock)
+                }
+
+                stats.rtcpPacketReceived(rtcpPacket, clock.instant().toEpochMilli())
+                "the rtt" {
+                    should("be updated correctly") {
+                        mostRecentPublishedRtt shouldBe(10.0.plusOrMinus(.1))
+                    }
+                }
+            }
+        }
+        "when a report block is received for which we can't find an SR" {
+            val reportBlock = mock<RtcpReportBlock> {
+                on { ssrc } doReturn 1234L
+                on { lastSrTimestamp } doReturn 100
+                on { delaySinceLastSr } doReturn 50.milliseconds.toDelaySinceLastSrFraction()
+            }
+
+            val rtcpPacket = mock<RtcpRrPacket> {
+                on { reportBlocks } doReturn listOf(reportBlock)
+            }
+
+            stats.rtcpPacketReceived(rtcpPacket, clock.instant().toEpochMilli())
+
+            "the rtt" {
+                should("not have been updated") {
+                    numRttUpdates shouldBe 0
+                }
+            }
+        }
+        "when a report block is received with no lastSrTimestamp" {
```

Comment:
It's not possible for a report block not to contain a lastSrTimestamp. (It's a fixed-offset field.) . This is related to my query about lastSrTimestamp being non-zero.

Reply:
what if none had been received? isn't that possible?